src/Pos/Communication/Limits.hs

```diff
@@ -0,0 +1,455 @@
+
+instance MessageLimitedPure UpdateVote where
+    msgLenLimit =
+        UpdateVote <$> msgLenLimit <*> msgLenLimit <*> msgLenLimit
+                   <*> msgLenLimit
+
+instance MessageLimitedPure SoftwareVersion where
+    msgLenLimit = SoftwareVersion <$> msgLenLimit <*> msgLenLimit
+
+instance MessageLimitedPure (DataMsg UpdateVote) where
+    msgLenLimit = DataMsg <$> msgLenLimit
+
+instance MessageLimitedPure (DataMsg (UpdateProposal, [UpdateVote])) where
+    msgLenLimit = DataMsg <$>
+        ((,) <$> msgLenLimit <*> vector updateVoteNumLimit)
+
+instance MessageLimitedPure TxMsgContents where
+    msgLenLimit = Limit Const.genesisMaxTxSize
+
+instance MessageLimitedPure (DataMsg TxMsgContents) where
+    msgLenLimit = DataMsg <$> msgLenLimit
+
+instance MessageLimitedPure a => MessageLimitedPure (Maybe a) where
+    msgLenLimit = Just <$> msgLenLimit + 1
+
+instance ( MessageLimitedPure a
+         , MessageLimitedPure b
+         )
+         => MessageLimitedPure (Either a b) where
+    msgLenLimit = 1 + max (Left <$> msgLenLimit) (Right <$> msgLenLimit)
+
+instance ( MessageLimitedPure a
+         , MessageLimitedPure b
+         )
+         => MessageLimitedPure (a, b) where
+    msgLenLimit = (,) <$> msgLenLimit <*> msgLenLimit
+
+instance ( MessageLimitedPure a
+         , MessageLimitedPure b
+         , MessageLimitedPure c
+         )
+         => MessageLimitedPure (a, b, c) where
+    msgLenLimit = (,,) <$> msgLenLimit <*> msgLenLimit <*> msgLenLimit
+
+instance MessageLimitedPure (Signature a) where
+    msgLenLimit = 64
+
+instance MessageLimitedPure PublicKey where
+    msgLenLimit = 32
+
+instance MessageLimitedPure Bool where
+    msgLenLimit = 1
+
+instance MessageLimitedPure a => MessageLimitedPure (AsBinary a) where
+    msgLenLimit = coerce (msgLenLimit :: Limit a) + 20
+
+instance MessageLimitedPure SecretProof where
+    msgLenLimit = 64
+
+instance MessageLimitedPure VssPublicKey where
+    msgLenLimit = 33
+
+instance MessageLimitedPure EncShare where
+    msgLenLimit = 101
+
+instance MessageLimitedPure Share where
+    msgLenLimit = 101 --4+33+64
+
+instance MessageLimitedPure PVSS.Commitment where
+    msgLenLimit = 33
+
+instance MessageLimitedPure PVSS.ExtraGen where
+    msgLenLimit = 33
+
+instance MessageLimitedPure (AbstractHash Blake2s_224 a) where
+    msgLenLimit = 28
+
+instance MessageLimitedPure (AbstractHash Blake2s_256 a) where
+    msgLenLimit = 32
+
+instance MessageLimitedPure BlockVersion where
+    msgLenLimit = 5
+
+instance MessageLimitedPure BlockVersionData where
+    msgLenLimit = 65
+
+instance MessageLimitedPure ApplicationName where
+    msgLenLimit = fromIntegral appNameLenLimit
+
+instance MessageLimitedPure Word32 where
+    msgLenLimit = 4
+
+instance MessageLimitedPure SystemTag where
+    msgLenLimit = 6
+
+instance MessageLimitedPure UpAttributes where
+#ifdef DEV_MODE
+    msgLenLimit = 0  -- real value
+#else
+    msgLenLimit = 10000  -- for backward compatibility
```

Comment:
I don't know what this value means, but I am almost sure it's irrelevant (it's some internal parameter of `ByteString` IIRC).
I think we should limit maximal proposal size and make it updateable using update system. In the same way as it's done (or maybe not done yet, I didn't follow) for transactions, headers, blocks (which have attributes too).